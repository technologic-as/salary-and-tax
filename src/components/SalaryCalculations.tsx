import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import React from 'react';
import { defineMessages, useIntl } from 'react-intl';
import { connect } from 'react-redux';
import { Section, TaxRow } from './Ui';
import { State } from '../reducers';

const messages = defineMessages({
  header: {
    id: 'salary.calculations.header',
    defaultMessage: 'Salary calculations',
  },
  turnover: { id: 'salary.calculations.turnover', defaultMessage: 'Turnover' },
  theCut: {
    id: 'salary.calculations.subcontractor.cut',
    defaultMessage: 'Subcontractor cut',
  },
  surplus: { id: 'salary.calculations.surplus', defaultMessage: 'Surplus' },
  employerFee: {
    id: 'salary.calculations.employer.fee',
    defaultMessage: 'Employer fee',
  },
  vacationSavings: {
    id: 'salary.calculations.vacation.savings',
    defaultMessage: 'Vacation savings',
  },
  pension: { id: 'salary.calculations.pension', defaultMessage: 'Pension' },
  income: { id: 'salary.calculations.income', defaultMessage: 'Income' },
});

type Props = {
  withoutPension: number;
  surplus: number;
  companyIncome: number;
  withoutVacationSavings: number;
  pension: number;
  employerFee: number;
  theCut: number;
  vacationSavings: number;
  turnover: number;
  withoutEmployerFee: number;
  withoutSurplus: number;
};

export const SalaryCalculationsComponent = ({
  turnover,
  companyIncome,
  theCut,
  employerFee,
  withoutEmployerFee,
  vacationSavings,
  withoutVacationSavings,
  pension,
  withoutPension,
  surplus,
  withoutSurplus,
}: Props) => {
  const { formatNumber, formatMessage } = useIntl();
  const formatCurrency = (amount: number) => formatNumber(amount, { style: 'currency', currency: 'NOK' });

  return (
    <Section header={formatMessage(messages.header)}>
      <Table>
        <TableBody>
          <TaxRow description={formatMessage(messages.turnover)} sum={formatCurrency(turnover)} />
          <TaxRow
            description={formatMessage(messages.theCut)}
            amount={formatCurrency(theCut)}
            sum={formatCurrency(companyIncome)}
            minus
          />
          <TaxRow
            description={formatMessage(messages.surplus)}
            amount={formatCurrency(surplus)}
            sum={formatCurrency(withoutSurplus)}
            minus
          />
          <TaxRow
            description={formatMessage(messages.employerFee)}
            amount={formatCurrency(employerFee)}
            sum={formatCurrency(withoutEmployerFee)}
            minus
          />
          <TaxRow
            description={formatMessage(messages.vacationSavings)}
            amount={formatCurrency(vacationSavings)}
            sum={formatCurrency(withoutVacationSavings)}
            minus
          />
          <TaxRow
            description={formatMessage(messages.pension)}
            amount={formatCurrency(pension)}
            sum={formatCurrency(withoutPension)}
            minus
          />
          <TaxRow description={formatMessage(messages.income)} sum={formatCurrency(withoutPension)} />
        </TableBody>
      </Table>
    </Section>
  );
};

const mapStateToProps = ({
  calculations: {
    salary: {
      turnover,
      companyIncome,
      theCut,
      employerFee,
      withoutEmployerFee,
      vacationSavings,
      withoutVacationSavings,
      pension,
      withoutPension,
      surplus,
      withoutSurplus,
    },
  },
}: State) => ({
  turnover,
  companyIncome,
  theCut,
  employerFee,
  withoutEmployerFee,
  vacationSavings,
  withoutVacationSavings,
  pension,
  withoutPension,
  surplus,
  withoutSurplus,
});

const mapDispatchToProps = () => ({});

// @ts-ignore
export const SalaryCalculations = connect(mapStateToProps, mapDispatchToProps)(SalaryCalculationsComponent);
export default SalaryCalculations;
