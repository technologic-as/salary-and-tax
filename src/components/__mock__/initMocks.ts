const mockFormatMessage = ({ defaultMessage }: { defaultMessage: string }) => defaultMessage;
const mockFormatNumber = (amount: number) =>
  new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(amount);

const mockIntl = () => ({
  formatMessage: mockFormatMessage,
  formatDate: jest.fn,
  formatTime: jest.fn,
  formatRelative: jest.fn,
  formatNumber: mockFormatNumber,
  formatPlural: jest.fn,
  formatHTMLMessage: jest.fn,
  now: jest.fn,
});

jest.mock('react-intl', () => ({
  useIntl: mockIntl,
  defineMessages: (msg: unknown) => msg,
}));

jest.mock('react-intl-redux', () => ({
  updateIntl: (a: unknown) => a,
}));

jest.mock('@material-ui/core/Table', () => 'Table');
jest.mock('@material-ui/core/TableBody', () => 'TableBody');
jest.mock('@material-ui/core/FormGroup', () => 'FormGroup');
jest.mock('@material-ui/core/FormControlLabel', () => 'FormControlLabel');
jest.mock('@material-ui/core/Radio', () => 'Radio');

export {};
