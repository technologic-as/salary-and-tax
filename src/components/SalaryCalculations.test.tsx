import mockComponent from 'identity-obj-proxy';
import React from 'react';
import renderer from 'react-test-renderer';
import './__mock__/initMocks';
import { SalaryCalculationsComponent } from './SalaryCalculations';

jest.mock('./Ui', () => mockComponent);
jest.mock('@material-ui/core/Table', () => 'Table');
jest.mock('@material-ui/core/TableBody', () => 'TableBody');

describe('SalaryCalculations', () => {
  it('should render', () => {
    const tree = renderer
      .create(
        <SalaryCalculationsComponent
          companyIncome={101010}
          employerFee={111111}
          pension={121212}
          theCut={131313}
          turnover={141414}
          vacationSavings={151515}
          withoutEmployerFee={161616}
          withoutPension={171717}
          withoutVacationSavings={181818}
          surplus={191919}
          withoutSurplus={202020}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
