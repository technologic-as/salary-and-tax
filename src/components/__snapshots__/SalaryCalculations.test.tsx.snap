// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalaryCalculations should render 1`] = `
<Section
  header="Salary calculations"
>
  <Table>
    <TableBody>
      <TaxRow
        description="Turnover"
        sum="$141,414.00"
      />
      <TaxRow
        amount="$131,313.00"
        description="Subcontractor cut"
        minus={true}
        sum="$101,010.00"
      />
      <TaxRow
        amount="$191,919.00"
        description="Surplus"
        minus={true}
        sum="$202,020.00"
      />
      <TaxRow
        amount="$111,111.00"
        description="Employer fee"
        minus={true}
        sum="$161,616.00"
      />
      <TaxRow
        amount="$151,515.00"
        description="Vacation savings"
        minus={true}
        sum="$181,818.00"
      />
      <TaxRow
        amount="$121,212.00"
        description="Pension"
        minus={true}
        sum="$171,717.00"
      />
      <TaxRow
        description="Income"
        sum="$171,717.00"
      />
    </TableBody>
  </Table>
</Section>
`;
